### PR TITLE
Implement GetBeaconState in the debug API

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

What does this PR do? Why is it needed?

This PR implements debug API's `GetBeaconState` method according to the spec: https://ethereum.github.io/eth2.0-APIs/#/Debug/getState

Elements:
- Implemented and tested a `ToProto` receiver method on the beacon state. There are tons of assertions, but I wanted make sure each field is correct because this function might end up being used in other places.
- Interface and mock for `StateFetcher`
- Since the fetcher does all the work in the gRPC function and it is tested thoroughly, the endpoint test only verifies if there are no obvious issues. 

Which issues(s) does this PR fix?

Part of #7510

Other notes for review
N/A

### CODE REVIEW

File: beacon-chain/rpc/statefetcher/fetcher.go [hunk 1]
```diff
@@ -16,7 +16,12 @@ import (
 	"github.com/prysmaticlabs/prysm/shared/bytesutil"
 )
 
-// StateFetcher is responsible for retrieving the BeaconState.
+// IStateFetcher is responsible for retrieving the BeaconState.
+type IStateFetcher interface {
```
Comment (rauljordan) [inline beacon-chain/rpc/statefetcher/fetcher.go:20]: Typically we avoid calling interface the same as a struct with just an `I` prefix, unless there is no better name. For example, we try to call interfaces by verbs, such as `StateFetcher`, `SlashingChecker`, etc. and the structs using a noun instead

File: beacon-chain/state/stateV0/state_trie.go [hunk 2]
```diff
@@ -204,6 +205,134 @@ func (b *BeaconState) HashTreeRoot(ctx context.Context) ([32]byte, error) {
 	return bytesutil.ToBytes32(b.merkleLayers[len(b.merkleLayers)-1][0]), nil
 }
 
+// ToProto returns a *v1.BeaconState representation of the state.
```
Comment (rauljordan) [inline beacon-chain/state/stateV0/state_trie.go:208]:
```suggestion
// ToProto returns a protobuf *v1.BeaconState representation of the state.
```

File: beacon-chain/state/stateV0/state_trie.go [hunk 3]
```diff
@@ -204,6 +205,134 @@ func (b *BeaconState) HashTreeRoot(ctx context.Context) ([32]byte, error) {
 	return bytesutil.ToBytes32(b.merkleLayers[len(b.merkleLayers)-1][0]), nil
 }
 
+// ToProto returns a protobuf *v1.BeaconState representation of the state.
+func (b *BeaconState) ToProto() *v1.BeaconState {
```
Comment (prestonvanloon) [inline beacon-chain/state/stateV0/state_trie.go:209]: Isn't this the same as CloneInnerState()?

Comment (rkapka) [reply beacon-chain/state/stateV0/state_trie.go:209]: Looks like it is the same. In addition `CloneInnerState` includes some additional safety measures. Thanks for noticing!

Comment (rkapka) [reply beacon-chain/state/stateV0/state_trie.go:209]: I looked more closely at the code today and it's actually not the same. `CloneInnerState` returns a non-ethereumapis state used by our old APIs.

**Review by rauljordan — APPROVED** [review]

**Pull request merged**